[SEC-1131] Extend SHA256.swift by functionality to calculate SHA256 checksum using a buffer

Simon-F-AMS opened the pull request:
… which is crucial for large files not to cause the app to run out of memory

- Add UnitTest ensuring this produces the same result as the existing SHA256 hashing via data
- Adjust TestData.swift by convenience url property in the process

Commit:
- Extend SHA256.swift by functionality to calculate SHA256 checksum using a buffer, which is crucial for large files not to cause the app to run out of memory
- Add UnitTest ensuring this produces the same result as the existing SHA256 hashing via data
- Adjust TestData.swift by conveinienve url property in the process

Simon-F-AMS commented:
Would prefer 2 approvals on the PR as this is security / crypto relevant stuff

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -0,0 +1,22 @@
+class SHA256Tests: XCTestCase {
+  func testSHA256Calculation_buffered__shouldProduceSameAsUnbuffered() {
+    // given
+    let testData = TestData.kvConnectEmailVerificationInvalidSignatureDigest
```
Would be good to use a random file of a size larger than the buffer

Simon-F-AMS commented:
As discussed:
- I will add a separate test file of sufficient file size (5 MB) to test the buffered sha256 calculation with multiple buffer fills
- I will add a SHA256 Test with a recalculated checksum from openssl and compare it to that

In a separate PR, I will
- Implement a variant of the SHA256 Hashing which uses CryptoKit, if available

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -20,4 +20,41 @@ public struct SHA256 {
+
+  // MARK: - Buffered SHA-256 Calculation
+
+  public static func digest(from file: URL, with bufferSize: Int = 1024 * 1024) -> Data? {
```
```suggestion
  public static func digest(file url: URL, withBufferSize bufferSize: Int = 1024 * 1024) throws -> Data {
```
Makes it clearer at call site: `SHA256.digest(file: URL(...), withBufferSize: 1_000_000)`

And with throwing it's better in case we need to handle specific error cases, especially since `FileHandle` is used.

hngry commented on Example/Tests/TestData.swift:
```diff
@@ -46,16 +46,21 @@ enum TestData: String {
     return data
   }
 
+  var url: URL {
```
Ideally this is in a separate PR but okay for now.

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -20,4 +20,41 @@ public struct SHA256 {
   }
+
+  // MARK: - Buffered SHA-256 Calculation
+
```
Should probably also add documentation info that bufferSize is in bytes

Commit:
- Add larger test file for the buffered SHA256 hash calculation test
- Add SHA256 test with pre calculated checksum via openssl
- Add large randomized test file in the process
- add better documentation to buffered sha256 calculation to explain that buffer is in bytes etc.

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -0,0 +1,34 @@
+    let bufferedDigest = try! SHA256.digest(file: testData.url)
+    // then
+    XCTAssertEqual(unbufferedDigest, bufferedDigest)
+  }
```
Don't like that the digest from file test is using the digest from data to produce the expected hash, it's not really a unit test at this point.

```swift
class SHA256Tests: XCTestCase {

  // Computed using `openssl dgst -sha256 -hex [file path]
  let expectedHash = "4b9ecf8a918594442fa56d14f2f8e975b258d134bf71c3a05c913b952d40af46"

    func testDigestData__shouldProduceExpectedHash() {
      // given
      let testData = TestData.largeTestData

      // when
      let digest = SHA256.digest(testData.data)

      // then
      XCTAssertEqual(digest, Data(hex: expectedHash))
    }
 
    func testDigestFile_whenTestDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
      // given
      let testData = TestData.largeTestData

      // when
      let digest = try! SHA256.digest(file: testData.url)

      // then
      XCTAssertEqual(digest, Data(hex: expectedHash))
    }
```
Can we do something like this?

hngry commented on Krypt/Source/SHA256.swift:
```diff
@@ -20,4 +20,45 @@ public struct SHA256 {
+  /// - Parameters:
+  ///   - url: The url to the file to calculate the SHA256 hash for
+  ///   - bufferSize: The size of the buffer to use in bytes, defaults to 1024 * 1024 bytes =  1MB
+  public static func digest(file url: URL, with bufferSize: Int = 1024 * 1024) throws -> Data {
```
```suggestion
  public static func digest(file url: URL, withBufferSize bufferSize: Int = 1024 * 1024) throws -> Data {
```
I think it still needs the `withBufferSize`, just "with" can be quite confusing with primitive types.

Commit:
- Adapt requested changes

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -0,0 +1,44 @@
+  // Computed using `openssl dgst -sha256 -hex [file path]
+  private let expectedHash = "4b9ecf8a918594442fa56d14f2f8e975b258d134bf71c3a05c913b952d40af46"
+
+  func testSAH256Calculation_nonBuffered__shouldProduceCorrectHash() {
```
Would appreciate following our naming convention: 
```
// ------------------------------------------------------------------------------------------------
//
// Test functions
//
// ------------------------------------------------------------------------------------------------
// Format:
func testWhatIsBeingTested_conditions_additionalCondition__expectedResult() {}

// Example:
func testSectionJournal_multipleSnapshots_noTodaySnapshot__shouldTakeLatestAsTodaysEntry() {}
```
```suggestion
  func testDigestData__shouldProduceExpectedHash() {
```

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -0,0 +1,44 @@
+    XCTAssertEqual(digest, Data(hex: expectedHash))
+  }
+
+  func testSHA256Calculation_buffered__shouldProduceCorrectHash() {
```
```suggestion
  func testDigestFile_whenDataLargeEnoughFor5BufferFills__shouldProduceExpectedHash() {
```

hngry commented on Example/Tests/SHA256Tests.swift:
```diff
@@ -0,0 +1,44 @@
+    XCTAssertEqual(bufferedDigest, Data(hex: expectedHash))
+  }
+
+  func testSHA256Calculation_bufferedWithCustomBufferSize__shouldProduceCorrectHash() {
```
```suggestion
  func testDigestFile_withCustomBufferSize__shouldProduceExpectedHash() {
```

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>

Simon-F-AMS committed:
Update Example/Tests/SHA256Tests.swift

Co-Authored-By: marko <marko@hungry.dev>
